opm/parser/eclipse/EclipseState/Schedule/Well.cpp

```diff
@@ -240,6 +248,209 @@ namespace Opm {
         m_completions->update(time_step, mutable_copy);
     }
 
+    void Well::addSegmentSet(size_t time_step, const SegmentSetPtr new_segmentset, const bool first_time) {
+
+        if (new_segmentset->lengthDepthType() == WellSegment::INC) {
+            addSegmentSetINC(time_step, new_segmentset, first_time);
+        } else if (new_segmentset->lengthDepthType() == WellSegment::ABS) {
+            addSegmentSetABS(time_step, new_segmentset, first_time);
+        } else {
+            throw std::runtime_error(" unknow length_depth_type in the new_segmentset");
+        }
+    }
+
+    void Well::addSegmentSetINC(size_t time_step, const SegmentSetPtr new_segmentset, const bool first_time) {
+        // The following code only applied when no loop exist.
+        // where loops exist, some modification should be made.
+        if (first_time) {
+            // update the information inside new_segmentset to be in ABS way
+            (*new_segmentset)[0]->length() = new_segmentset->lengthTopSegment();
```

Comment:
Hmmm - looks like the use of references was intentional? I have not looked detailed into this yet - but be prepared; you are going to have a really hard time convincing me that this way of manipulating the internal state of the segment objects is kosher.